File: Day24/minimum-score-after-removals-on-a-tree.js
// Time: O(N^2)  ::  Space: O(N) ==> For both approaches

class Solution {
    calc(a, b, c) {
        return Math.max(a, b, c) - Math.min(a, b, c);
    }

    minimumScore(nums, edges) {
        const n = nums.length;
        const adj = Array.from({ length: n }, () => []);
        for (const [u, v] of edges) {
            adj[u].push(v);
            adj[v].push(u);
        }

        const tin = Array(n).fill(0), tout = Array(n).fill(0), sumXor = Array(n).fill(0);
        let timer = 0;

        const dfs = (x, par) => {
            tin[x] = timer++;
            sumXor[x] = nums[x];
            for (const y of adj[x]) {
                if (y === par) continue;
                dfs(y, x);
                sumXor[x] ^= sumXor[y];
            }
            tout[x] = timer;
        };

        dfs(0, -1);

        let res = Infinity;
        for (let u = 1; u < n; ++u) {
            for (let v = u + 1; v < n; ++v) {
                if (tin[v] > tin[u] && tin[v] < tout[u]) {
                    res = Math.min(res, this.calc(sumXor[0] ^ sumXor[u], sumXor[u] ^ sumXor[v], sumXor[v]));
                } else if (tin[u] > tin[v] && tin[u] < tout[v]) {
                    res = Math.min(res, this.calc(sumXor[0] ^ sumXor[v], sumXor[v] ^ sumXor[u], sumXor[u]));
                } else {
                    res = Math.min(res, this.calc(sumXor[0] ^ sumXor[u] ^ sumXor[v], sumXor[u], sumXor[v]));
                }
            }
        }

        return res;
    }
}





// Two-pass DFS Aproach
class Solution {
    constructor() {
        this.res = Infinity;
        this.sumXor = 0;
    }

    calc(a, b, c) {
        return Math.max(a, b, c) - Math.min(a, b, c);
    }

    dfs2(node, parent, firstXor, anc, nums, adj) {
        let secondXor = nums[node];
        for (const child of adj[node]) {
            if (child === parent) continue;
            secondXor ^= this.dfs2(child, node, firstXor, anc, nums, adj);
        }
        if (parent !== anc) {
            this.res = Math.min(this.res, this.calc(firstXor, secondXor, this.sumXor ^ firstXor ^ secondXor));
        }
        return secondXor;
    }

    dfs(node, parent, nums, adj) {
        let firstXor = nums[node];
        for (const child of adj[node]) {
            if (child === parent) continue;
            firstXor ^= this.dfs(child, node, nums, adj);
        }
        for (const child of adj[node]) {
            if (child !== parent) {
                this.dfs2(child, node, firstXor, node, nums, adj);
            }
        }
        return firstXor;
    }

    minimumScore(nums, edges) {
        const n = nums.length;
        const adj = Array.from({ length: n }, () => []);
        for (const [u, v] of edges) {
            adj[u].push(v);
            adj[v].push(u);
        }

        this.sumXor = nums.reduce((acc, val) => acc ^ val, 0);
        this.res = Infinity;
        this.dfs(0, -1, nums, adj);
        return this.res;
    }
}

File: Day24/minimum-score-after-removals-on-a-tree.cc
// Time: O(N^2)  ::  Space: O(N) ==> For both approaches

class Solution {
public:
    int calc(int a, int b, int c) {
        return max({a, b, c}) - min({a, b, c});
    }

    int minimumScore(vector<int>& nums, vector<vector<int>>& edges) {
        int n = nums.size();
        vector<vector<int>> adj(n);
        for (auto& edge : edges) {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        vector<int> sumXor(n), tin(n), tout(n);
        int timer = 0;

        function<void(int, int)> dfs = [&](int node, int parent) {
            tin[node] = timer++;
            sumXor[node] = nums[node];
            for (int child : adj[node]) {
                if (child == parent) continue;
                dfs(child, node);
                sumXor[node] ^= sumXor[child];
            }
            tout[node] = timer;
        };

        dfs(0, -1);

        int res = INT_MAX;
        for (int u = 1; u < n; ++u) {
            for (int v = u + 1; v < n; ++v) {
                if (tin[v] > tin[u] && tin[v] < tout[u]) {
                    res = min(res, calc(sumXor[0] ^ sumXor[u], sumXor[u] ^ sumXor[v], sumXor[v]));
                } else if (tin[u] > tin[v] && tin[u] < tout[v]) {
                    res = min(res, calc(sumXor[0] ^ sumXor[v], sumXor[v] ^ sumXor[u], sumXor[u]));
                } else {
                    res = min(res, calc(sumXor[0] ^ sumXor[u] ^ sumXor[v], sumXor[u], sumXor[v]));
                }
            }
        }

        return res;
    }
};



// Two-pass DFS Aproach
class Solution {
public:
    int res = INT_MAX, sumXor = 0;

    int calc(int a, int b, int c) {
        return max({a, b, c}) - min({a, b, c});
    }

    int dfs2(int node, int parent, int firstXor, int anc, vector<int>& nums, vector<vector<int>>& adj) {
        int secondXor = nums[node];
        for (int child : adj[node]) {
            if (child == parent) continue;
            secondXor ^= dfs2(child, node, firstXor, anc, nums, adj);
        }
        if (parent != anc) {
            res = min(res, calc(firstXor, secondXor, sumXor ^ firstXor ^ secondXor));
        }
        return secondXor;
    }

    int dfs(int node, int parent, vector<int>& nums, vector<vector<int>>& adj) {
        int firstXor = nums[node];
        for (int child : adj[node]) {
            if (child == parent) continue;
            firstXor ^= dfs(child, node, nums, adj);
        }
        for (int child : adj[node]) {
            if (child != parent) {
                dfs2(child, node, firstXor, node, nums, adj);
            }
        }
        return firstXor;
    }
